Zend/zend_compile.c

```diff
@@ -6637,6 +6641,497 @@ static void zend_compile_pipe(znode *result, zend_ast *ast)
 	zend_compile_expr(result, fcall_ast);
 }
 
+typedef struct {
+	bool inside_or_pattern;
+	zend_stack labels;
+	uint32_t num_bindings;
+	zend_stack bindings;
+} zend_pm_context;
+
+typedef struct {
+	uint32_t cv;
+	znode value;
+} zend_pm_binding;
+
+static void zend_compile_pattern(zend_ast *ast, znode *result, znode *expr_node, bool consume_expr, uint32_t false_opnum, zend_pm_context *context);
+static zend_type zend_compile_single_typename(zend_ast *ast);
+
+static void zend_pm_context_init(zend_pm_context *context)
+{
+	zend_stack_init(&context->labels, sizeof(uint32_t));
+	/* Avoid offsets 0 and 1. */
+	uint32_t dummy = 0;
+	zend_stack_push(&context->labels, &dummy);
+	zend_stack_push(&context->labels, &dummy);
+
+	zend_stack_init(&context->bindings, sizeof(zend_pm_binding));
+}
+
+static void zend_pm_context_free(zend_pm_context *context)
+{
+	zend_stack_destroy(&context->labels);
+	zend_stack_destroy(&context->bindings);
+}
+
+static uint32_t zend_pm_label_create(zend_pm_context *context)
+{
+	uint32_t dummy = 0;
+	return -zend_stack_push(&context->labels, &dummy);
```

Comment:
The negation together with the `uint32_t` return looks very odd. Perhaps a comment is helpful (or is this obvious and I just don't know the API well?).